### PR TITLE
Update Shallow Snapshot flows to support remote path type & hash algo

### LINKED ISSUE #12987: [Remote Store] Update shallow snapshot flows to support new path types and hash algorithm
### Is your feature request related to a problem? Please describe

With optimised prefix pattern for remote store path (as mentioned in https://github.com/opensearch-project/OpenSearch/issues/12567), we need to ensure that we are able to allow indexes to use the optimised remote store path type and at the same time be able to resolve the existing prefix pattern (Fixed remote store path type) during snapshots and restores.

### Describe the solution you'd like
We have updated Snapshots and it's restore after introducing shallow snapshots with remote store. Due to the changes in the blob store path for different type of data, it becomes important that we keep track of the same in shallow snapshot metadata since the snapshot can be restored outside of the life of an index (meaning that the cluster manager holds no information about the index itself). I propose the solution where we will update the shallow snapshot metadata file that will keep track of the remote store type and hash algorithm so that we can reuse this information during times of 1. restore 2. clone 3. cleaning up remote store data on account of snapshot deletion when there are no more lock files present for the same.

### Related component

Storage:Performance

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ashking94**
### Description
This PR is a sub task of feature request https://github.com/opensearch-project/OpenSearch/issues/12567 and being tracked in meta issue - https://github.com/opensearch-project/OpenSearch/issues/12589.

In this PR, we have wired the RemoteStorePathStrategy to Shallow Snapshots. We have written new tests and modified the existing tests as required. We handle following senarios -
1. Snapshots and restores
2. Snapshot clone
3. Lock file deletion during snapshot deletion
4. Remote store cleanup after every snapshot deletion

### Related Issues
Resolves #12987

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- ~[ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~
- ~[ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java [hunk 1]
```diff
@@ -237,6 +251,10 @@ public static RemoteStoreShardShallowCopySnapshot fromXContent(XContentParser pa
                     remoteStoreRepository = parser.text();
                 } else if (PARSE_REPOSITORY_BASE_PATH.match(currentFieldName, parser.getDeprecationHandler())) {
                     repositoryBasePath = parser.text();
+                } else if (PARSE_PATH_TYPE.match(currentFieldName, parser.getDeprecationHandler())) {
+                    pathType = PathType.fromCode(parser.intValue());
+                } else if (PARSE_PATH_HASH_ALGORITHM.match(currentFieldName, parser.getDeprecationHandler())) {
+                    pathHashAlgorithm = PathHashAlgorithm.fromCode(parser.intValue());
                 } else {
                     throw new OpenSearchParseException("unknown parameter [{}]", currentFieldName);
                 }
```
[the remark this answers is not in the document]
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:260]: I did think about it. It should not be a problem during version upgrade. We are introducing a check that will ensure that thew new prefix approach is enabled only after min node version is above 2.14 where this feature will be released. Since we disallow restoring snapshots taken on higher opensearch version onto lower opensearch version, I dont think we should have any issues.

**Comment by ashking94** [comment]
> ❕ Gradle check result for [2ef85af](https://github.com/opensearch-project/OpenSearch/commit/2ef85afa574e95356746f19503ea93207553335c): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/36371/)
> 
> * **TEST FAILURES:**
> 
> ```
>       3 org.opensearch.cluster.MinimumClusterManagerNodesIT.testThreeNodesNoClusterManagerBlock
>       1 org.opensearch.index.IndexServiceTests.testAsyncTranslogTrimTaskOnClosedIndex
>       1 org.opensearch.http.SearchRestCancellationIT.testAutomaticCancellationDuringQueryPhase
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

Flaky tests - https://github.com/opensearch-project/OpenSearch/issues/10006, https://github.com/opensearch-project/OpenSearch/issues/11547 & https://github.com/opensearch-project/OpenSearch/issues/5426

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java [hunk 2]
```diff
@@ -327,16 +329,22 @@ public void testRemoteStoreCustomDataOnIndexCreationAndRestore() {
         validatePathType(indexName2, PathType.HASHED_PREFIX, PathHashAlgorithm.FNV_1A);
 
         // Validating that custom data has not changed for indexes which were created before the cluster setting got updated
-        validatePathType(indexName1, PathType.FIXED, PathHashAlgorithm.FNV_1A);
+        validatePathType(indexName1, PathType.FIXED);
```
Comment (harishbhakuni) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java:332]: Can we add one test to restore on the same index, basically `create idx -> take snapshot -> update cluster settings -> close idx -> restore idx from snapshot -> validate that path type is still PathType.FIXED`.

File: server/src/main/java/org/opensearch/index/IndexSettings.java [hunk 3]
```diff
@@ -1910,12 +1911,11 @@ public void setDocIdFuzzySetFalsePositiveProbability(double docIdFuzzySetFalsePo
...
-        if (remoteCustomData != null
-            && remoteCustomData.containsKey(PathType.NAME)
-            && remoteCustomData.containsKey(PathHashAlgorithm.NAME)) {
+        if (remoteCustomData != null && remoteCustomData.containsKey(PathType.NAME)) {
             PathType pathType = PathType.parseString(remoteCustomData.get(PathType.NAME));
-            PathHashAlgorithm pathHashAlgorithm = PathHashAlgorithm.parseString(remoteCustomData.get(PathHashAlgorithm.NAME));
-            return new RemoteStorePathStrategy(pathType, pathHashAlgorithm);
+            String hashAlgoStr = remoteCustomData.get(PathHashAlgorithm.NAME);
+            PathHashAlgorithm hashAlgorithm = Objects.nonNull(hashAlgoStr) ? PathHashAlgorithm.parseString(hashAlgoStr) : null;
+            return new RemoteStorePathStrategy(pathType, hashAlgorithm);
```
Comment (harishbhakuni) [inline server/src/main/java/org/opensearch/index/IndexSettings.java:1918]: if pathType is set and hashAlgorithm is not set and passed as null, will it use FIXED path strategy?

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java [hunk 4]
```diff
@@ -101,9 +112,21 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
+        }
         return builder;
     }
 
+    /**
+     * This is kept only to avoid breaking changes.
+     */
+    @Deprecated
     public RemoteStoreShardShallowCopySnapshot(
         String snapshot,
```
Comment (harishbhakuni) [inline server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:131]: which breaking changes? this will be called from BlobStoreRepository which we must be updating already.

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java [hunk 5]
```diff
@@ -449,4 +526,55 @@ public IndexShardSnapshotStatus getIndexShardSnapshotStatus() {
...
+            && this.startTime == that.startTime
+            && this.time == that.time
+            && this.totalFileCount == that.totalFileCount
+            && this.totalSize == that.totalSize
+            && this.primaryTerm == that.primaryTerm
+            && this.commitGeneration == that.commitGeneration
+            && Objects.equals(this.remoteStoreRepository, that.remoteStoreRepository)
+            && Objects.equals(this.repositoryBasePath, that.repositoryBasePath)
+            && Objects.equals(this.indexUUID, that.indexUUID)
+            && Objects.equals(this.fileNames, that.fileNames)
```
Comment (harishbhakuni) [inline server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:555]: fileNames is a list of strings, we should instead use equals method of list which will return true if the elements of both lists are same. also, are we using this method currently?

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java [hunk 6]
```diff
@@ -237,6 +297,10 @@ public static RemoteStoreShardShallowCopySnapshot fromXContent(XContentParser pa
                     remoteStoreRepository = parser.text();
                 } else if (PARSE_REPOSITORY_BASE_PATH.match(currentFieldName, parser.getDeprecationHandler())) {
                     repositoryBasePath = parser.text();
+                } else if (PARSE_PATH_TYPE.match(currentFieldName, parser.getDeprecationHandler())) {
+                    pathType = PathType.fromCode(parser.intValue());
+                } else if (PARSE_PATH_HASH_ALGORITHM.match(currentFieldName, parser.getDeprecationHandler())) {
+                    pathHashAlgorithm = PathHashAlgorithm.fromCode(parser.intValue());
                 } else {
                     throw new OpenSearchParseException("unknown parameter [{}]", currentFieldName);
                 }
```
Comment (harishbhakuni) [inline server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:306]: should we remove this condition for future? so that even if the code is old, it would not fail as we are verifying the parameters at the end anyways.

File: server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java [hunk 7]
```diff
@@ -2707,7 +2704,9 @@ public void snapshotRemoteStoreIndexShard(
                         store.indexSettings().getUUID(),
                         store.indexSettings().getRemoteStoreRepository(),
                         this.basePath().toString(),
-                        fileNames
+                        fileNames,
+                        pathStrategy.getType(),
+                        pathStrategy.getHashAlgorithm()
```
Comment (harishbhakuni) [inline server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java:2709]: while writing to the file we are passing pathType and hashAlgorithm, however while reading we are using `getRemoteStorePathStrategy` method. would it be better if we directly pass RemoteStorePathStrategy instance in constructor and internally use these attributes while parsing or creating xcontent object?

**Review by harishbhakuni — CHANGES_REQUESTED** [review]

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java [hunk 8]
```diff
@@ -327,16 +329,22 @@ public void testRemoteStoreCustomDataOnIndexCreationAndRestore() {
         validatePathType(indexName2, PathType.HASHED_PREFIX, PathHashAlgorithm.FNV_1A);
```
Comment (harishbhakuni) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java:329]: as this value will be in index metadata, do we need to override it during restore based on cluster defaults? otherwise i think it will be picked from snapshot metadata.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java (hunk 2 above)
Comment (ashking94) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java:332]: `testRemoteStoreCustomDataOnIndexCreationAndRestore`already covers this case. I will adding more ITs as part of https://github.com/opensearch-project/OpenSearch/issues/12766 after https://github.com/opensearch-project/OpenSearch/issues/13074 (All changes are wired up).

File: server/src/main/java/org/opensearch/index/IndexSettings.java (hunk 3 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/IndexSettings.java:1918]: We had an assertion earlier, but **now i have replaced with an exception**. This should never happen that pathType is hash based and there is no path hash algorithm. Such cases are bugs and should be fixed.

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java (hunk 4 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:131]: You can refer the PR check `Detect Breaking Changes / detect-breaking-change (pull_request) `.

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java (hunk 5 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:555]: `Objects.equals()` below:
```
    public static boolean equals(Object a, Object b) {
        return (a == b) || (a != null && a.equals(b));
    }
```

File: server/src/main/java/org/opensearch/index/IndexSettings.java (hunk 3 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/IndexSettings.java:1918]: should we memoize`RemoteStorePathStrategy` in index settings ?

File: server/src/main/java/org/opensearch/index/remote/RemoteStoreEnums.java [hunk 9]
```diff
@@ -167,6 +199,33 @@ long hash(PathInput pathInput) {
             }
         };
 
+        private final int code;
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/index/remote/RemoteStoreEnums.java:202]: would prefer to use `string` instead of `int` as it will be less error prone and more explicit , when we add more `PathHashAlgorithm`  .

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java (hunk 4 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:131]: Is this exposed to plugins in any way ? Looks like this is an internal class and won't be a breaking change . We should mark it as `InternalApi` .

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java [hunk 10]
```diff
@@ -41,8 +46,10 @@ public class RemoteStoreShardShallowCopySnapshot implements ToXContentFragment,
     private final String repositoryBasePath;
     private final String indexUUID;
     private final List<String> fileNames;
+    private final PathType pathType;
+    private final PathHashAlgorithm pathHashAlgorithm;
 
-    static final String DEFAULT_VERSION = "1";
+    static final String DEFAULT_VERSION = "2";
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:52]: where are we using it ?

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java (hunk 8 above)
Comment (ashking94) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java:329]: The index metadata needs to be overridden to ensure that the restored index uses the latest path strategy. This has been done currently already.

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java (hunk 6 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:306]: This is a very important check to ensure that we do not restore a shallow snapshot on a version that does not understand new fields. Removing this check can lead to inconsistent restore.

File: server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java (hunk 7 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java:2709]: Sure, will do this.

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java [hunk 11]
```diff
@@ -206,8 +206,9 @@ public MetadataCreateIndexService(
 
         // Task is onboarded for throttling, it will get retried from associated TransportClusterManagerNodeAction.
         createIndexTaskKey = clusterService.registerClusterManagerTask(ClusterManagerTaskKeys.CREATE_INDEX_KEY, true);
+        Supplier<Version> minNodeVersionSupplier = () -> clusterService.state().nodes().getMinNodeVersion();
```
[the remark this answers is not in the document]
Comment (ashking94) [reply server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:209]: In my first draft, I was passing the clusterService, then thought it is better to only pass the supplier to limit access to this important service.

File: server/src/main/java/org/opensearch/index/remote/RemoteStoreEnums.java (hunk 9 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/remote/RemoteStoreEnums.java:202]: Using int is a standard practise since it maps elements of an enum one to one also limiting the size of the object. This has been followed in OpenSearch as well.

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java (hunk 10 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:52]: This is being used in `isValidRemotePathConfiguration` method.

File: server/src/main/java/org/opensearch/index/IndexSettings.java (hunk 3 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/IndexSettings.java:1918]: Done

File: server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java (hunk 7 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java:2709]: I attempted this change, but this has downsides of multiple checks (if/else) conditions to be added and also inhibits the capability to validate path type and hash algorithm against the version. For that reason, I am keeping this unchanged.

File: server/src/main/java/org/opensearch/index/remote/RemoteStorePathStrategyResolver.java [hunk 12]
```diff
@@ -22,13 +25,22 @@ public class RemoteStorePathStrategyResolver {
...
         type = clusterSettings.get(IndicesService.CLUSTER_REMOTE_STORE_PATH_PREFIX_TYPE_SETTING);
         clusterSettings.addSettingsUpdateConsumer(IndicesService.CLUSTER_REMOTE_STORE_PATH_PREFIX_TYPE_SETTING, this::setType);
     }
 
     public RemoteStorePathStrategy get() {
-        return new RemoteStorePathStrategy(type, PathHashAlgorithm.FNV_1A);
+        PathType pathType;
+        PathHashAlgorithm pathHashAlgorithm;
+        // Min node version check ensures that we are enabling the new prefix type only when all the nodes understand it.
+        pathType = Version.CURRENT.compareTo(minNodeVersionSupplier.get()) <= 0 ? type : PathType.FIXED;
```
[the remark this answers is not in the document]
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/remote/RemoteStorePathStrategyResolver.java:40]: After the backport of this PR to 2.x branch, I will udpate current to 2.14 and backport the same.

File: server/src/main/java/org/opensearch/index/remote/RemoteStoreEnums.java (hunk 9 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/remote/RemoteStoreEnums.java:202]: Made the change.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java (hunk 2 above)
Comment (harishbhakuni) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java:332]: no, we are using different index there during restore. we need to close the same index and restore it from snapshot. verified that it will work fine from code perspective, but it would be good to have a test, maybe in the above followup PRs.

Comment (ashking94) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteRestoreSnapshotIT.java:332]: Ohk, we are restoring but as a new index. I get your point, will try to put this in the same PR.

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java (hunk 4 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:131]: This class has to be marked public api due to below reference in Repository class -
https://github.com/opensearch-project/OpenSearch/blob/9ab31f55e88cc2d023cad6e60a15a0e0e87dac38/server/src/main/java/org/opensearch/repositories/Repository.java#L369-L375
Not changing anything.

**Comment by ashking94** [comment]
> org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testIndexReopenClose

Flaky test - https://github.com/opensearch-project/OpenSearch/issues/13084

**Review by harishbhakuni — APPROVED** [review]
changes lgtm

File: server/src/main/java/org/opensearch/index/IndexSettings.java (hunk 3 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/IndexSettings.java:1921]: If we are using the hashed prefix, then index metadata would be updated before the first shard starts on remote node.  So i don't think we would need changes . LMK if i am missing something

File: server/src/main/java/org/opensearch/index/remote/RemoteStoreEnums.java [hunk 13]
```diff
@@ -112,6 +118,40 @@ boolean requiresHashAlgorithm() {
...
+            CODE_TO_ENUM = unmodifiableMap(codeToStatus);
+        }
+
+        /**
+         * Turn a status code into a {@link PathType}.
+         */
+        public static PathType fromCode(int code) {
+            return CODE_TO_ENUM.get(code);
+        }
+
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/index/remote/RemoteStoreEnums.java:154]: instead of `fromCode`, `toCode` , can we implement `writeTo(StreamOutput out)` and `readFrom(StreamInput in)` methods. We can refer to `AllocationStatus` for a sample implementation  ?

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java [hunk 14]
```diff
@@ -118,32 +141,60 @@ public RemoteStoreShardShallowCopySnapshot(
...
-        this.totalFileCount = totalFileCount;
-        this.totalSize = totalSize;
-        this.indexUUID = indexUUID;
-        this.remoteStoreRepository = remoteStoreRepository;
-        this.repositoryBasePath = repositoryBasePath;
-        this.fileNames = fileNames;
     }
 
-    private RemoteStoreShardShallowCopySnapshot(
+    public RemoteStoreShardShallowCopySnapshot(
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:197]: why we are changing it ? If it was private , nobody would be calling it from outside of this package . Then we wouldn't need two versions ,right ?

File: server/src/test/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshotTests.java [hunk 15]
```diff
@@ -94,30 +163,76 @@ public void testFromXContent() throws IOException {
...
+        );
+        try (XContentParser parser = createParser(JsonXContent.jsonXContent, xContent)) {
+            RemoteStoreShardShallowCopySnapshot actualShardShallowCopySnapshot = RemoteStoreShardShallowCopySnapshot.fromXContent(parser);
+            assert Objects.equals(expectedShardShallowCopySnapshot, actualShardShallowCopySnapshot);
         }
     }
 
     public void testFromXContentInvalid() throws IOException {
-        final int iters = scaledRandomIntBetween(1, 10);
+        final int iters = 14;
```
Comment (gbbafna) [inline server/src/test/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshotTests.java:235]: why this change ?

File: server/src/main/java/org/opensearch/index/IndexSettings.java (hunk 3 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/IndexSettings.java:1921]: > If we are using the hashed prefix, then index metadata would be updated before the first shard starts on remote node.
Not required for this case. If we are updating index metadata at the end, then we do need it.

File: server/src/main/java/org/opensearch/index/remote/RemoteStoreEnums.java (hunk 13 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/remote/RemoteStoreEnums.java:154]: We dont have StreamInput or StreamOutput here.

File: server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java (hunk 14 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshot.java:197]: Earlier the ctor that was exposed did not have the version field. We need ctor with version for writing tests for both the versions. However, we dont need this to be public. I have changed it to default access.

File: server/src/test/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshotTests.java (hunk 15 above)
Comment (ashking94) [reply server/src/test/java/org/opensearch/index/snapshots/blobstore/RemoteStoreShardShallowCopySnapshotTests.java:235]: This will test out all possible cases of failures and success each test run. Otherwise earlier we were running few tests only.

**Review by gbbafna — APPROVED** [review]

**Comment by gbbafna** [comment]
`org.opensearch.discovery.ClusterDisruptionIT.testAckedIndexing` failed . Retrying gradle build .

**Comment by ashking94** [comment]
> testConcurrentDecommissionAction

Flaky test - https://github.com/opensearch-project/OpenSearch/issues/12197
